### PR TITLE
apimachinery: fix UnsafeGuessKindToResource for some classes of words

### LINKED ISSUE #1082: Pluralization discrepancy in UnsafeGuessKindToResource still exists
Expected:

gateway-> gateways
Got:
gateway-> gatewaies

Is there any way to affect this mistaken guessing?

### PR DESCRIPTION
**Pull request by michaelbeaumont**
#### What type of PR is this?

/kind bug

#### What this PR does / why we need it:

Fixes guesses that `apimachinery` makes about resource names from a GVK.

#### Which issue(s) this PR fixes:

Fixes https://github.com/kubernetes/client-go/issues/1082

#### Does this PR introduce a user-facing change?

```release-note
NONE
```

### CODE REVIEW

**Comment by apelisse** [comment]
/ok-to-test
/lgtm

Not sure if this breaks compatibility with anything.

**Comment by AlexanderYastrebov** [comment]
The https://github.com/kubernetes/code-generator [uses](https://github.com/kubernetes/code-generator/blob/e4dc4396b33332454a61312abb46d57833d3554c/cmd/client-gen/generators/client_generator.go#L42) https://pkg.go.dev/k8s.io/gengo/namer#NewAllLowercasePluralNamer so another option might be to use it here or make client-go use it instead of `UnsafeGuessKindToResource` that is documented as broken.

**Comment by evankanderson** [comment]
FWIW, this affects https://gateway-api.sigs.k8s.io/, which is sort of annoying (spent about 2-3 hours debugging why a test framework using `.Add` wasn't finding my Gateway).

👍  on this change from me.

**Comment by pmalek** [comment]
Also hit this when using Gateway API's objects with dynamic fake client: k8s.io/client-go/dynamic/fake (just as @evankanderson described it).

Any (even rough) ETA for this? Any workarounds before this gets merged?

**Comment by pmalek** [comment]
Any chance that we can get someone else to look at this and review it (as @deads2k and @apelisse seem not be able to pick this up)?

**Comment by apelisse** [comment]
Why would you throw me under the bus? I've lgtm'd 6 months ago! 😄

**Comment by pmalek** [comment]
@apelisse sorry 😅 Didn't notice that (and the fact that still @deads2k is assigned as reviewer: so I guess we need lgtm from him). My bad.

**Review by pmalek — APPROVED** [review]

**Comment by lavalamp** [comment]
I think we might have this code locked because changing it could break existing things and pluralizing english is hopeless unless we can call out to GPT-3 anyway.

There *should* be fields to define both in CRDs to avoid this code being invoked, if that's not the case please let us know.

File: staging/src/k8s.io/apimachinery/pkg/api/meta/restmapper.go
Comment (liggitt) [inline staging/src/k8s.io/apimachinery/pkg/api/meta/restmapper.go:140]: /hold

This isn't a compatible change to make to this function. Anything using this function to auto-generate plural names from singular names would break when re-generating an existing object that this change impacts.

I think we should remove all non-test in-tree calls to this function, mark it deprecated, file issues to public callers we find, and consider removing it in a future release.

If we want a *new* function that takes a singular name and pluralizes it that has this expanded behavior from the start (with a signature like `AttemptToPluralize(singular string) (plural string)`), I guess I wouldn't object, but that doesn't need to live in k8s.io/apimachinery

Comment (evankanderson) [reply staging/src/k8s.io/apimachinery/pkg/api/meta/restmapper.go:140]: I believe that the use that bit me _was_ in test coffee without an apiserver loaded. I can try to dig out the code in question; I believe it tries to pluralize "Gateway" as "Gatewaies", which was very confusing when I finally debugged it.

Removing this function would also be acceptable to me, but it's currently a trap for anyone testing the Gateway API.

Comment (pmalek) [reply staging/src/k8s.io/apimachinery/pkg/api/meta/restmapper.go:140]: @evankanderson Can you dig that test code up? I'm positive I've also encountered this issue recently but can't find the relevant test code that would manifest the problem itself (other than calling `UnsafeGuessKindToResource` with Gateway type).

Comment (evankanderson) [reply staging/src/k8s.io/apimachinery/pkg/api/meta/restmapper.go:140]: I believe that we were getting hung up in `tracker.Add` from this code:

https://github.com/kubernetes/client-go/blob/master/testing/fixture.go#L341

I also see some questionable usage in https://github.com/knative/pkg/blob/main/resolver/addressable_resolver.go#L167 and https://github.com/knative/pkg/blob/main/kref/knative_reference.go#L53, but the usage of `ObjectTracker` from `client-go/testing` seems most likely to cause problems, because I'm guessing that package is used many places.

You can see my work-around here:

https://github.com/knative-sandbox/net-gateway-api/blob/main/pkg/reconciler/ingress/ingress_test.go#L404

**Comment by liggitt** [comment]
/close

Changing this function is not the right direction. I would be in favor of marking as deprecated and removing all in-tree calls to it, and adding the ability for fake clients to be told the accurate plural/singular mappings if needed.

**Comment by pmalek** [comment]
@liggitt I'd like to move forward with this somehow but I'm having hard time understanding your proposal.

E.g. how would you go about removing in-tree calls of this? There are multiple places where it's used in user facing code e.g.

https://github.com/kubernetes/kubernetes/blob/694698ca3801fe5aa258d58c1e27ac7f88c29d35/staging/src/k8s.io/apimachinery/pkg/api/meta/restmapper.go#L99-L113

Would you suggest to mark `func (m *DefaultRESTMapper) Add(kind schema.GroupVersionKind, scope RESTScope)` as deprecated as well due to its usage of `UnsafeGuessKindToResource` and suggest users to use `AddSpecific()` instead?

**Comment by pmalek** [comment]
It seems that using `k8s.io/gengo/namer`'s `NewAllLowercasePluralNamer()` would work:

```
func (m *DefaultRESTMapper) Add(kind schema.GroupVersionKind, scope RESTScope) {
	n := namer.NewAllLowercasePluralNamer(nil)
	t := types.Ref(kind.GroupVersion().String(), kind.Kind)
	plural := kind.GroupVersion().WithResource(n.Name(t))
	singular := kind.GroupVersion().WithResource(strings.ToLower(kind.Kind))

	m.AddSpecific(kind, plural, singular, scope)
}
```

If this would be accepted and is perceived as the direction that we can move forward with, I can submit a PR replacing usages of `UnsafeGuessKindToResource` with the above.

**Comment by liggitt** [comment]
yes, mark deprecated and mark callers that rely on it as deprecated as well, pointing people to AddSpecific

**Comment by pmalek** [comment]
@liggitt I've pushed https://github.com/kubernetes/kubernetes/pull/119924 to try to address the above PTAL.

**Pull request closed without merging**
